File: core/r/tests/testthat/test-two-group.R
# Integration check on the two-group recipe: the stats path must produce the
# forced test, the right group sizes, and a significant p on well-separated data.
test_that("recipe_two_group_compare runs the forced Welch t-test end to end", {
  df <- data.frame(
    group = rep(c("ctrl", "treat"), each = 8),
    value = c(10, 11, 12, 13, 14, 10.5, 11.5, 12.5,
              20, 21, 22, 23, 24, 20.5, 21.5, 22.5)
  )
  opt <- list(recipe = "two_group_compare", data = "toy.csv",
              x = "group", y = "value", test = "welch_t", paired = FALSE,
              p_adjust = "none")
  spec <- spec_from_opt(opt)

  res <- recipe_two_group_compare(df, spec)

  expect_named(res, c("plot", "stats", "test_meta", "resolved", "methods",
                      "build_script", "df_used", "clean_steps", "qc"),
               ignore.order = TRUE)
  expect_s3_class(res$plot, "ggplot")
  expect_equal(res$stats$test, "Welch two-sample t-test")
  expect_equal(res$stats$n1, 8)
  expect_equal(res$stats$n2, 8)
  expect_true(res$stats$p_value < 0.01)
  expect_setequal(c(res$stats$group1, res$stats$group2), c("ctrl", "treat"))
})

test_that("recipe_two_group_compare rejects a grouping column that is not binary", {
  df <- data.frame(group = rep(c("a", "b", "c"), each = 3), value = 1:9)
  opt <- list(recipe = "two_group_compare", data = "toy.csv",
              x = "group", y = "value", test = "auto")
  spec <- spec_from_opt(opt)
  expect_error(recipe_two_group_compare(df, spec), "exactly 2 groups")
})

test_that("the Mann-Whitney branch reports a signed effect size and Levene's p", {
  set.seed(7)
  df <- data.frame(
    group = rep(c("control", "treated"), each = 12),
    value = c(rlnorm(12, 1, 0.4), rlnorm(12, 2, 0.4)),
    stringsAsFactors = FALSE
  )
  opt <- list(recipe = "two_group_compare", data = "toy.csv",
              x = "group", y = "value", test = "wilcoxon", paired = FALSE,
              p_adjust = "none")
  spec <- spec_from_opt(opt)

  out <- recipe_two_group_compare(df, spec)

  expect_equal(out$stats$test, "Mann-Whitney U test")
  expect_equal(out$stats$effect_size_name, "rank-biserial r")
  expect_false(is.na(out$stats$effect_size))
  expect_lt(out$stats$effect_size, 0)   # control ranks below treated
  expect_gte(out$stats$effect_size, -1)
  expect_equal(out$test_meta$assumptions$equal_variance$test, "Levene")
  expect_false(is.na(out$stats$var_equal_p))
})

File: core/python/tests/test_two_group.py
from types import SimpleNamespace

import pandas as pd
import pytest

from pubready.recipes.two_group_compare import _resolve, recipe_two_group_compare
from pubready.spec import spec_from_args


def test_resolve_picks_test_from_assumptions():
    assert _resolve("auto", True, True, False)["label"] == "Student's two-sample t-test"
    assert _resolve("auto", True, False, False)["label"] == "Welch two-sample t-test"
    assert _resolve("auto", False, True, False)["label"] == "Mann-Whitney U test"
    assert _resolve("welch_t", False, True, False)["var_equal"] is False
    assert _resolve("t", True, True, True)["label"] == "paired t-test"
    assert _resolve("wilcoxon", False, True, True)["label"] == "Wilcoxon signed-rank test"
    with pytest.raises(ValueError, match="unknown test method"):
        _resolve("nonsense", True, True, False)


def test_recipe_two_group_runs_forced_welch_end_to_end():
    df = pd.DataFrame({
        "group": ["ctrl"] * 8 + ["treat"] * 8,
        "value": [10, 11, 12, 13, 14, 10.5, 11.5, 12.5,
                  20, 21, 22, 23, 24, 20.5, 21.5, 22.5],
    })
    a = SimpleNamespace(recipe="two_group_compare", data="toy.csv",
                        x="group", y="value", test="welch_t")
    spec = spec_from_args(a)

    res = recipe_two_group_compare(df, spec)

    row = res["stats"].iloc[0]
    assert row["test"] == "Welch two-sample t-test"
    assert row["n1"] == 8 and row["n2"] == 8
    assert row["p_value"] < 0.01
    assert {row["group1"], row["group2"]} == {"ctrl", "treat"}


def test_recipe_two_group_rejects_non_binary_group():
    df = pd.DataFrame({"group": ["a", "b", "c"] * 3, "value": range(9)})
    a = SimpleNamespace(recipe="two_group_compare", data="toy.csv",
                        x="group", y="value", test="auto")
    spec = spec_from_args(a)
    with pytest.raises(ValueError, match="exactly 2 groups"):
        recipe_two_group_compare(df, spec)
